Improve the processing of `enum` in `ocean.text.convert.Formatter`

MichaelKim20 opened the pull request:
I improved the processing of `enum` in `ocean.text.convert.Formatter`.

Relates to https://github.com/bpfkorea/agora/issues/725

MichaelKim20 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -381,9 +381,22 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
+    // Print enum type and member name
     else static if (is (T V == enum))
-        handle!(V)(v, f, sf, se);
+    {
```
The base type of the enum only works normally with an integer or string.
This causes errors in other base types.
``` D
    import std.stdio;
    enum FooB : real {
        one = 1,
        two = 1.41421,
        three = 1.73205
    }
    FooB fb = FooB.one;
    writeln(format("{}", fb));
```

AndrejMitrovic replied:
For enums where you can't do a `switch` it's possible to do a linear search like this:

```D
void handleNonSwitchableEnum()
{
    static foreach (idx, member; EnumAllMembers!T)
    {
        if (v == member)
            sf(T.stringof ~ "." ~ __traits(identifier, EnumAllMembers!T[idx]));
            return;
    }

    sf("cast(" ~ T.stringof ~ ") ");
    handle!(V)(v, f, sf, se);
}

handleNonSwitchableEnum();
```

But it's not very efficient because of `O(n)` comparisons.

I could think of another solution: to hash the value and reduce it to a `ulong` which can then be used in a switch/case. But this would require a lot more code. Also it's not currently possible with our hashing routines in `Hash.d` because they cannot run at compile-time (CTFE): `source/agora/common/Hash.d(118,28): Error: crypto_generichash_init cannot be interpreted at compile time, because it has no available source code`.

Geod24 replied:
What is the error you are seeing ?

MichaelKim20 replied:
``` D
sw: switch (v)
```
I was thinking about when other types of data came because v is only an integer and string.

Geod24 replied:
Right, good point. Honestly non-integral / string enums are so rare that we can leave it out for the time being. Just make it fall back to the underlying type when that happens.

MichaelKim20 replied:
That's a good idea.

Geod24 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -381,9 +381,22 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
+    {
+        sw: switch (v)
+        {
+            static foreach (idx, member; EnumAllMembers!T)
```
As discussed, remove the `static` here

MichaelKim20 replied:
That's right. I remember.

Geod24 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -911,3 +924,95 @@ private struct FormatInfo
+
+    template MemberIdentifier(string ident)
+    {
+        static if (ident == "MakeAlias")
```
What is this ?

MichaelKim20 replied:
It means when the ident happens to be "MakeAlias."

MichaelKim20 replied:
I'll remove this.

Geod24 commented on src/ocean/text/convert/Formatter_test.d:
```diff
@@ -424,9 +424,19 @@ unittest
+    fa = cast(FooA)"beta";
+    test("FooA.b" == format("{}", fa));
+    fa = cast(FooA)"gamma";
+    test("cast(FooA) gamma" == format("{}", fa));
```
Please also add test for non-allocation (see `testNoAlloc` examples)

MichaelKim20 replied:
That's a good suggestion.

Geod24 commented on src/ocean/text/convert/Formatter_test.d:
```diff
@@ -423,10 +429,34 @@ unittest
-    test("42" == format("{}", f));
-    f = cast(Foo)36;
-    test("36" == format("{}", f));
+    string[] outputs;
+    outputs.length = 3;
+    assumeSafeAppend(outputs);
+
+    Foo[] foo_inputs;
+    foo_inputs.length = 3;
+    assumeSafeAppend(foo_inputs);
+    testNoAlloc(foo_inputs.fill(Foo.FooBar, cast(Foo)1, cast(Foo)36));
+    testNoAlloc(outputs.fill("Foo.FooBar", "Foo.B", "cast(Foo) 36"));
+    foreach (i, ref item; foo_inputs)
+        test!("==")(outputs[i], format("{}", item));
```
```suggestion
    char[256] buffer;
    immutable Foo[] foo_inputs = [ Foo.FooBar, cast(Foo)1, cast(Foo)36 ];
    immutable string[] foo_outputs = [ "Foo.FooBar", "Foo.B", "cast(Foo) 36" ];
    foreach (i, ref item; foo_inputs)
        testNoAlloc(assert(sformat(buff, "{}", item) == output[i]));
```

We want to test that `sformat` does not allocate. The test currently test that `fill` does not allocate. Why ?

MichaelKim20 replied:
> We want to test that `sformat` does not allocate. The test currently test that `fill` does not allocate. Why ?

I was mistaken.

MichaelKim20 replied:
sformat(buffer, "{}", item)   =>
result is "immutable(Foo).FooBar"

MichaelKim20 committed:
Implement `enum` formatting in `ocean.text.convert.Formatter`

Added processing for `enum` to function handle in `ocean.text.convert.Formatter`.

AndrejMitrovic commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -381,9 +381,44 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
+            foreach (member; __traits(allMembers, T))
+            {
+                case mixin("T.", member):
+                    sf(T.stringof ~ "." ~ member);
```
`~` will allocate a new string. I suggest doing something like this instead:

```D
sf(T.stringof);
sf(".");
sf(member);
```

And please do this for all other calls to `sf` below where `~` was used.

Geod24 replied:
Note that, since they are all CT-known values, they get const-folded. But since it's not always obvious, that's why I asked for the `testNoAlloc` check. In any case your suggestion makes it explicit so 👍

MichaelKim20 replied:
I revised it to your suggestion. Thank you for making such a good suggestion.

AndrejMitrovic reviewed (changes requested):
Just one small issue with allocations.

MichaelKim20 commented:
Ready for review ~

Geod24 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -381,9 +381,52 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
-        handle!(V)(v, f, sf, se);
+
+    // Print enum member name - If `switch statement` is available
+    else static if (is(T V == enum) && canSwitchOn!(V))
```
```suggestion
    else static if (is(T V == enum) && canSwitchOn!(T))
```

MukeunKim replied:
I knew V was an element.
So I used elements of the element.

Geod24 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -381,9 +381,52 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
+    }
+
+    // Print enum member name - If `switch statement` is not available
+    else static if (is(T E == enum) && !canSwitchOn!(E))
```
```suggestion
    else static if (is(T E == enum))
```

Geod24 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -381,9 +381,52 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
+                sf(T.stringof);
+                sf(".");
+                sf(member);
+                done = true;
```
Instead of using `done`, just `return`

Geod24 commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -911,3 +949,16 @@ private struct FormatInfo
 }
+
+/// Verify that the 'switch statement' is available
+template canSwitchOn (T)
```
```suggestion
private template canSwitchOn (T)
```

Geod24 reviewed (approved):
LGTM, just one minor change

Geod24 commented:
Upstreamed as https://github.com/sociomantic-tsunami/ocean/pull/824
